### ISSUE TITLE
Incorrect detection of the types of keys and values in a dictionary.

### ISSUE THREAD

**Issue by phi-friday** [microsoft/pyright#9058]
**Describe the bug**
In the code below, the key value and value should be detected as string and any, respectively.
I've noticed that I don't get an error until `1.1.379`, and then I get an error starting with `1.1.380`.

##### Definition of `Dataset`
```python
@attr.define()
class Dataset(os.PathLike, BaseDataset):
    uri: str = attr.field(
        converter=_sanitize_uri,
        validator=[attr.validators.min_len(1), attr.validators.max_len(3000)],
    )
    extra: dict[str, Any] | None = None
```

**Code or Screenshots**

```python
from __future__ import annotations

from typing import TYPE_CHECKING, Any

from airflow.datasets import Dataset
from airflow.utils.context import Context

if TYPE_CHECKING:
    from collections.abc import Mapping


def f(conf: dict[str, Any] | Context) -> Mapping[str, Mapping[str, Any]]: ...


def g(values: Mapping[str, Mapping[str, Any]] | Context) -> Mapping[str, Dataset]:
    if isinstance(values, Context):  # pyright: ignore[reportArgumentType]
        values = f(values)

    return {key: Dataset(uri=key, extra=dict(value)) for key, value in values.items()}

```

**VS Code extension or command-line**
```bash
❯ uvx --from "pyright==1.1.379" pyright test.py --pythonpath=.venv/bin/python
WARNING: there is a new pyright version available (v1.1.379 -> v1.1.381).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

0 errors, 0 warnings, 0 informations 
❯ uvx --from "pyright==1.1.380" pyright test.py --pythonpath=.venv/bin/python
WARNING: there is a new pyright version available (v1.1.380 -> v1.1.381).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

/Users/***/git/python/repo/kma_airflow_server/test.py
  /Users/***/git/python/repo/kma_airflow_server/test.py:19:41 - error: No overloads for "__init__" match the provided arguments (reportCallIssue)
  /Users/***/git/python/repo/kma_airflow_server/test.py:19:41 - error: Argument of type "dict[bytes, bytes]" cannot be assigned to parameter "extra" of type "dict[str, Any] | None" in function "__init__"
    Type "dict[bytes, bytes]" is not assignable to type "dict[str, Any] | None"
      "dict[bytes, bytes]" is not assignable to "dict[str, Any]"
        Type parameter "_KT@dict" is invariant, but "bytes" is not the same as "str"
      "dict[bytes, bytes]" is not assignable to "None" (reportArgumentType)
  /Users/***/git/python/repo/kma_airflow_server/test.py:19:46 - error: Argument of type "Mapping[str, Any] | object" cannot be assigned to parameter "iterable" of type "Iterable[list[bytes]]" in function "__init__"
    Type "Mapping[str, Any] | object" is not assignable to type "Iterable[list[bytes]]"
      "object" is incompatible with protocol "Iterable[list[bytes]]"
        "__iter__" is not present (reportArgumentType)
3 errors, 0 warnings, 0 informations 
❯ uvx --from "pyright==1.1.381" pyright test.py --pythonpath=.venv/bin/python
/Users/***/git/python/repo/kma_airflow_server/test.py
  /Users/***/git/python/repo/kma_airflow_server/test.py:19:41 - error: No overloads for "__init__" match the provided arguments (reportCallIssue)
  /Users/***/git/python/repo/kma_airflow_server/test.py:19:41 - error: Argument of type "dict[bytes, bytes]" cannot be assigned to parameter "extra" of type "dict[str, Any] | None" in function "__init__"
    Type "dict[bytes, bytes]" is not assignable to type "dict[str, Any] | None"
      "dict[bytes, bytes]" is not assignable to "dict[str, Any]"
        Type parameter "_KT@dict" is invariant, but "bytes" is not the same as "str"
      "dict[bytes, bytes]" is not assignable to "None" (reportArgumentType)
  /Users/***/git/python/repo/kma_airflow_server/test.py:19:46 - error: Argument of type "Mapping[str, Any] | object" cannot be assigned to parameter "iterable" of type "Iterable[list[bytes]]" in function "__init__"
    Type "Mapping[str, Any] | object" is not assignable to type "Iterable[list[bytes]]"
      "object" is incompatible with protocol "Iterable[list[bytes]]"
        "__iter__" is not present (reportArgumentType)
3 errors, 0 warnings, 0 informations 
```

**Comment by erictraut** [comment]
Your code samples reference symbols that are not defined. Could you please provide a minimal, self-contained sample that demonstrates the problem? Preferably a single piece of code that I can paste into a blank code file that demonstrates the problem. I tried to piece together a sample from your code, but I wasn't able to repro the errors you're seeing, so I assume that I'm missing something. If you are relying on third-party libraries, please specify which libraries and versions I need to install to repro the issue.

**Comment by phi-friday** [comment]
> Your code samples reference symbols that are not defined. Could you please provide a minimal, self-contained sample that demonstrates the problem? Preferably a single piece of code that I can paste into a blank code file that demonstrates the problem. I tried to piece together a sample from your code, but I wasn't able to repro the errors you're seeing, so I assume that I'm missing something. If you are relying on third-party libraries, please specify which libraries and versions I need to install to repro the issue.

Sorry, I forgot to add dependencies.

#### python version
```bash
❯ uv run python -VV
Python 3.12.2 (main, Feb 25 2024, 03:55:42) [Clang 17.0.6 ]
```
#### dependencies
```bash
❯ uv pip freeze | grep apache-airflow==
apache-airflow==2.10.2
```
#### pyrightconfig.json
```json
{
    "typeCheckingMode": "strict",
    "pythonVersion": "3.12",
    "pythonPlatform": "Linux",
    "reportMissingTypeStubs": false,
    "reportInconsistentConstructor": false,
    "reportUnknownArgumentType": false,
    "reportOverlappingOverload": false,
    "reportUnknownVariableType": false,
    "reportUnknownMemberType": false,
    "reportUnnecessaryIsInstance": false,
    "reportUnnecessaryComparison": false,
    "reportDeprecated": false,
    "reportPrivateUsage": false,
    "reportUnnecessaryTypeIgnoreComment": true,
    "reportImplicitOverride": true,
    "reportIncompatibleVariableOverride": false,
    "enableExperimentalFeatures": true,
    "reportTypedDictNotRequiredAccess": false, // airflow Context
    "stubPath": "./typings"
}
```

**Comment by phi-friday** [comment]
The class `airflow.utils.context.Context` is actually a `MutableMapping`, but in the stub file it's defined as a `TypedDict`. I think that's what's causing the problem, is that correct?

The code below is the code I was able to see the error without `airflow`.

```python
from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any, TypedDict


@dataclass
class Dataset:
    uri: str
    extra: dict[str, Any] | None = None


class Context(TypedDict): ...


def f(conf: Context) -> dict[str, dict[str, Any]]: ...


def g(values: Mapping[str, Mapping[str, Any]] | Context) -> Mapping[str, Dataset]:
    if isinstance(values, Context):  # pyright: ignore[reportArgumentType]
        values = f(values)

    return {key: Dataset(uri=key, extra=dict(value)) for key, value in values.items()}
```

**Comment by erictraut** [comment]
I don't see a bug here. Pyright's type analysis looks correct to me. Can you explain why you think the behavior is incorrect?

Your code contains a bug, which you've suppressed with a `# pyright: ignore` comment. I'm not sure what the intent of this code is, but using a `TypedDict` in an `isinstance` call will not work. If you attempt to run this code, it will crash.

**Comment by phi-friday** [comment]
> I don't see a bug here. Pyright's type analysis looks correct to me. Can you explain why you think the behavior is incorrect?
> 
> Your code contains a bug, which you've suppressed with a `# pyright: ignore` comment. I'm not sure what the intent of this code is, but using a `TypedDict` in an `isinstance` call will not work. If you attempt to run this code, it will crash.

Yes, using `TypedDict` in `isinstance` will result in an error.
`Context` appears as `TypedDict` in the stub file, but at runtime it is a class, not `TypedDict`.
I think that's what's causing the problem.

**Comment by phi-friday** [comment]
The code below seems to represent the current situation.

```python
from __future__ import annotations

from collections.abc import Mapping, MutableMapping
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, TypedDict


@dataclass
class Dataset:
    uri: str
    extra: dict[str, Any] | None = None


if TYPE_CHECKING:

    class Context(TypedDict): ...

else:

    class Context(MutableMapping): ...


def f(conf: Context) -> dict[str, dict[str, Any]]: ...


def g(values: Mapping[str, Mapping[str, Any]] | Context) -> Mapping[str, Dataset]:
    if isinstance(values, Context):  # pyright: ignore[reportArgumentType]
        values = f(values)

    return {key: Dataset(uri=key, extra=dict(value)) for key, value in values.items()}
```

**Comment by erictraut** [comment]
Yeah, that's the problem. If you use the real definition of `Context` (the one that is consistent with the class used at runtime), the code type checks without any issues. The problem is that you're effectively "lying" to pyright about the real type here.

I think pyright is working correctly here, so I'm going to close the issue.

**Comment by phi-friday** [comment]
@erictraut 

I have two questions Could you please answer them further?

1. why did `1.1.379` not have the problem?
> I know it's not appropriate to ask about `pylance` here, but...
2. why is `pylance==2024.9.2` not throwing an error?(`pylance==2024.9.2` is using `pyright==1.1.381`)

**Comment by phi-friday** [comment]
The error also occurs when the `check()` function is used instead of `pyright: ignore`.

```python
from __future__ import annotations

from collections.abc import Mapping, MutableMapping
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, TypedDict

from typing_extensions import TypeIs


@dataclass
class Dataset:
    uri: str
    extra: dict[str, Any] | None = None


if TYPE_CHECKING:

    class Context(TypedDict): ...

else:

    class Context(MutableMapping): ...


def check(value: Any) -> TypeIs[Context]: ...


def f(conf: Context) -> dict[str, dict[str, Any]]: ...


def g(values: Mapping[str, Mapping[str, Any]] | Context) -> Mapping[str, Dataset]:
    if check(values):
        values = f(values)

    return {key: Dataset(uri=key, extra=dict(value)) for key, value in values.items()}

```

**Comment by erictraut** [comment]
I recently fixed some bugs and inconsistencies in the `isinstance` narrowing logic. Part of the intent behind these changes was to make the behavior if `isinstance` narrowing consistent with `TypeIs`. That probably explains why you are seeing a difference in behavior.

Pylance doesn't use the most recent version of pyright. It's typically one or two weeks behind. That probably explains why you're seeing different behavior here.

**Comment by phi-friday** [comment]
> I recently fixed some bugs and inconsistencies in the `isinstance` narrowing logic. Part of the intent behind these changes was to make the behavior if `isinstance` narrowing consistent with `TypeIs`. That probably explains why you are seeing a difference in behavior.
> 
> Pylance doesn't use the most recent version of pyright. It's typically one or two weeks behind. That probably explains why you're seeing different behavior here.

Yes, I was wrong about that.
It looks like `2024.9.101` uses `1.1.381`, 
but `2024.9.2` is still using the older version (`2024.9.2` was released after `2024.9.101`, so I thought `2024.9.2` had the latest version.)
Thanks for your answer.
